torchqdynamics/mesolve.py

```diff
@@ -126,8 +111,26 @@ def forward(self, t, dt, rho):
             M0 @ rho @ M0.adjoint() + rho_ + 0.5 * dt *
             (M1s @ rho_.unsqueeze(0) @ M1s.adjoint()).sum(dim=0)
         )
-        rho = rho / rho.trace().real
-        return rho
+        return rho / btrace(rho)
 
     def forward_adjoint(self, t, dt, phi):
         raise NotImplementedError
+
+
+def btrace(input):
+    """Compute the batched trace of a tensor over its last two dimensions, and return a
+    tensor of the same number of dimensions as input."""
+    return input.diagonal(dim1=-1, dim2=-2).sum(-1)[(..., ) + (None, ) * 2]
```

Comment:
I suggest the use of the beautiful `einsum` 😉 
```python
def btrace(x):
    # x: (..., d, d)
    # -> (..., d)
    return torch.einsum('...ii', x)
```

Reply:
I have this tendency to avoid `einsum` because I don't trust its performance. I have seen examples where it's very slow compared to normal functions.

Reply:
But maybe it's only in numpy, I'm not sure

Reply:
Ok it seems fixed in https://github.com/pytorch/pytorch/issues/32591. I will change for einsum then.